README.md
# ChatGPT-API-unity

A client library of [ChatGPT chat completion API](https://platform.openai.com/docs/api-reference/chat/create) for Unity.

See also official [document](https://platform.openai.com/docs/guides/chat) and [API reference](https://platform.openai.com/docs/api-reference/chat).

## How to import by Unity Package Manager

Add following dependencies to your `/Packages/mainfest.json`.

```json
{
  "dependencies": {
    "com.mochineko.chatgpt-api": "https://github.com/mochi-neko/ChatGPT-API-unity.git?path=/Assets/Mochineko/ChatGPT_API#0.7.3",
    ...
  }
}
```

## How to use chat completion by ChatGPT API

1. Generate API key on [OpenAI](https://platform.openai.com/account/api-keys). (Take care your API key, this is a secret information then you should not open.)
2. You can specify chat model. (Available models are defined by `Model`.)
3. Create an instance of `ChatCompletionAPIConnection` with API key and chat model. (This instance memorizes old messages in session.)
4. You can set system message (prompt) to instruct assistant with your situation by constructor of `ChatCompletionAPIConnection`.
5. Input user message and call `ChatCompletionAPIConnection.CompleteChatAsync()`.
6. Response message is in `ChatCompletionResponseBody.ResultMessage` (= `ChatCompletionResponseBody.Choices[0].Message.Content`).

An essential sample code with [UniTask](https://github.com/Cysharp/UniTask) is as follows:

```csharp
#nullable enable
using System;
using System.Threading;
using Cysharp.Threading.Tasks;
using Mochineko.ChatGPT_API.Memories;
using UnityEngine;

namespace Mochineko.ChatGPT_API.Samples
{
    /// <summary>
    /// A sample component to complete chat by ChatGPT API on Unity.
    /// </summary>
    public sealed class ChatCompletionSample : MonoBehaviour
    {
        /// <summary>
        /// API key generated by OpenAPI.
        /// </summary>
        [SerializeField] private string apiKey = string.Empty;

        /// <summary>
        /// System message to instruct assistant.
        /// </summary>
        [SerializeField, TextArea] private string systemMessage = string.Empty;

        /// <summary>
        /// Message sent to ChatGPT API.
        /// </summary>
        [SerializeField, TextArea] private string message = string.Empty;

        /// <summary>
        /// Max number of chat memory of queue.
        /// </summary>
        [SerializeField] private int maxMemoryCount = 20;

        private ChatCompletionAPIConnection? connection;
        private IChatMemory? memory;

        private void Start()
        {
            // API Key must be set.
            if (string.IsNullOrEmpty(apiKey))
            {
                Debug.LogError("OpenAI API key must be set.");
                return;
            }

            memory = new FiniteQueueChatMemory(maxMemoryCount);

            // Create instance of ChatGPTConnection with specifying chat model.
            connection = new ChatCompletionAPIConnection(
                apiKey,
                memory,
                systemMessage);
        }

        [ContextMenu(nameof(SendChat))]
        public void SendChat()
        {
            SendChatAsync(this.GetCancellationTokenOnDestroy()).Forget();
        }
        
        [ContextMenu(nameof(ClearChatMemory))]
        public void ClearChatMemory()
        {
            memory?.ClearAllMessages();
        }
        
        private async UniTask SendChatAsync(CancellationToken cancellationToken)
        {
            // Validations
            if (connection == null)
            {
                Debug.LogError($"[ChatGPT_API.Samples] Connection is null.");
                return;
            }

            if (string.IsNullOrEmpty(message))
            {
                Debug.LogError($"[ChatGPT_API.Samples] Chat content is empty.");
                return;
            }

            ChatCompletionResponseBody response;
            try
            {
                await UniTask.SwitchToThreadPool();
                
                // Create message by ChatGPT chat completion API.
                response = await connection.CompleteChatAsync(
                        message,
                        cancellationToken);
            }
            catch (Exception e)
            {
                // Exceptions should be caught.
                Debug.LogException(e);
                return;
            }

            await UniTask.SwitchToMainThread(cancellationToken);

            // Log chat completion result.
            Debug.Log($"[ChatGPT_API.Samples] Result:\n{response.ResultMessage}");
        }
    }
}
```

See also [Sample](./Assets/Mochineko/ChatGPT_API.Samples/ChatCompletionSample.cs).

## How to use chat completion by ChatGPT API more resilient

See `RelentChatCompletionAPIConnection` and `RelentChatCompletionSample`
 using [Relent](https://github.com/mochi-neko/Relent).

You can use API with explicit error handling, retry, timeout, bulkhead, and so on.

```json
{
  "dependencies": {
    "com.mochineko.chatgpt-api.relent": "https://github.com/mochi-neko/ChatGPT-API-unity.git?path=/Assets/Mochineko/ChatGPT_API.Relent#0.7.3",
    "com.mochineko.chatgpt-api": "https://github.com/mochi-neko/ChatGPT-API-unity.git?path=/Assets/Mochineko/ChatGPT_API#0.7.3",
    "com.mochineko.relent": "https://github.com/mochi-neko/Relent.git?path=/Assets/Mochineko/Relent#0.2.0",
    "com.cysharp.unitask": "https://github.com/Cysharp/UniTask.git?path=src/UniTask/Assets/Plugins/UniTask",
    ...
  }
}
```

## How to calculate token length of text in local

You can calculate token length of text
 by `TiktokenSharp.Tiktoken.Encode(string)` as follows:

```csharp
using TiktokenSharp;

private int CalculateTokenLength()
{
    string text = "A text that you want to calculate token length.";

    // Specify model name.
    Tiktoken tiktoken = TikToken.EncodingForModel("gpt-3.5-turbo");
    
    // Encoding is tokenizing.
    int[] tokens = tikToken.Encode(text);
    
    return tokens.Length;
}
```

If you want to calculate on Unity editor,
 please use `ItemuMenu > Mochineko > TiktokenEditor` window.

## How to customize chat memories

You can use customized memories of chat by implementing `IChatMemory` interface.

Presets are available:

- `FiniteQueueChatMemory`
  - A queue that has max number of messages.
- `FiniteQueueWithFixedPromptsChatMemory`
  - A queue that has max number of user/assistant messages and free number of prompts (system messages).
- `FiniteTokenLengthQueueChatMemory`
  - A queue that has max number of token lenght of all messages.
- `FiniteTokenLengthQueueWithFixedPromptsChatMemory`
  - A queue that has max number of token lenght of user/assistant messages and free number of prompts (system messages).

## How to stream response

See [streaming sample](./Assets/Mochineko/ChatGPT_API.Samples/ChatCompletionAsStreamSample.cs).

You can await foreach as follows:

```csharp
var builder = new StringBuilder();
 // Receive enumerable from ChatGPT chat completion API.
var enumerable = await connection.CompleteChatAsStreamAsync(
    message,
    cancellationToken);

await foreach (var chunk in enumerable.WithCancellation(cancellationToken))
{
    // First chunk has only "role" element.
    if (chunk.Choices[0].Delta.Content is null)
    {
        Debug.Log($"[ChatGPT_API.Samples] Role:{chunk.Choices[0].Delta.Role}.");
        continue;
    }
    
    var delta = chunk.Choices[0].Delta.Content;
    builder.Append(delta);
    Debug.Log($"[ChatGPT_API.Samples] Delta:{delta}, Current:{builder}");
}

// Log chat completion result.
Debug.Log($"[ChatGPT_API.Samples] Completed: \n{builder}");
```
.

## How to use function calling

1. Define function with JSON schema.
2. Specify function by request parameters.
3. Call chat completion API.
4. Use `result.Choices[0].Message.FunctionCall` 

See `FunctionCalling()` in [test code](./Assets/Mochineko/ChatGPT_API.Tests/ChatCompletionAPIConnectionTest.cs).

## Changelog

See [CHANGELOG](./CHANGELOG.md).

## 3rd Party Notices

See [NOTICE](./NOTICE.md).

## License

Licensed under the [MIT](./LICENSE) license.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.7.3] - 2023-07-21

### Fixed

- Fix default path to save PBE file in TikTokenSharp.
- Fill recorded message content with empty string because messages in request parameter must have content field.

## [0.7.2] - 2023-07-12

### Fixed

- Fix dependencies in `package.json`.

## [0.7.1] - 2023-07-05

### Fixed

- Fix request to add `HttpCompletionOption.ResponseHeadersRead` to Relent streaming API.

## [0.7.0] - 2023-07-05

### Added

- Add Streaming API sample.

### Changed

- Change streaming API to use `IAsyncEnumerable`.
- Change to `UniTask` from `Task` in Relent API.

## [0.6.0] - 2023-06-14

### Added

- Add 16k context turbo model option.
- Add function calling support.
- Add optional `name` property to `Message`.
- Add verbose log option.

## Changed

- Update chat models at 2023-06-13.
- Add dependencies to `package.json`.

## [0.5.0] - 2023-04-08

### Changed

- Simplify package structure.
- Update Relent version to 0.2.0.

## [0.4.0] - 2023-03-25

### Added

- Add extensions of `IChatMemory`.

### Changed

- Improve interface of `IChatMemory` for async operation.
- Be `IChatMemory` implementations thread-safe.

### Fixed

- Fix capacity of `FiniteQueueChatMemory`.

## [0.3.1] - 2023-03-25

### Changed

- Improve accessibility of `Messaage`.

## [0.3.0] - 2023-03-23

### Added

- Add `TiktokenSharp` that is a tokenizer to calculate token length of text in local.

## [0.2.2] - 2023-03-21

### Added

- Add option to specify instance of `HttpClient`.
- Add option to receive response as `Stream`.

### Changed

- Improve null check of response content.

## [0.2.1] - 2023-03-20

### Added

- Add GPT-4 all models.

## [0.2.0] - 2023-03-20

### Added

- Add memory management of chat completion API by `IChatMemory`.
- Add [resilient error handling implementation](https://github.com/mochi-neko/ChatGPT-API-unity/blob/main/Assets/Mochineko/ChatGPT_API.Relent/RelentChatCompletionAPIConnection.cs) of chat completion API by [Relent](https://github.com/mochi-neko/Relent).
- Add GPT-4 model enum.

### Changed

- Change name and arguments of completion method.

### Fixed

- Fix deserialization error of `"usage"` in response body.

## [0.1.1] - 2023-03-06

### Added

- Add optional request parameters.

### Fixed

- Fix error handling of API response.

## [0.1.0] - 2023-03-04

### Added

- Implement ChatGPT chat completion API bindings to C#.
- Implement sample component of ChatGPT API.


Assets/Mochineko/ChatGPT_API/Model.cs
#nullable enable
using System;

namespace Mochineko.ChatGPT_API
{
    /// <summary>
    /// Chat model.
    /// See also https://platform.openai.com/docs/models/model-endpoint-compatibility.
    /// </summary>
    public enum Model : byte
    {
        /// <summary>
        /// "gpt-3.5-turbo" latest GPT3.5 turbo model.
        /// </summary>
        Turbo,
        /// <summary>
        /// "gpt-3.5-turbo-0301" fixed GPT3.5 turbo model at 03/01/2023.
        /// </summary>
        [Obsolete("Use Turbo0613 instead of Turbo0301 until 09/13/2023.")]
        Turbo0301,
        /// <summary>
        /// "gpt-3.5-turbo-0613" fixed GPT3.5 turbo model at 06/13/2023.
        /// </summary>
        Turbo0613,
        /// <summary>
        /// "gpt-3.5-turbo-16k" latest GPT3.5 turbo model for 16k tokens.
        /// </summary>
        Turbo16K,
        /// <summary>
        /// "gpt-3.5-turbo-16k-0613" fixed GPT3.5 turbo model for 16k tokens at 06/13/2023.
        /// </summary>
        Turbo16K0613,
        /// <summary>
        /// "gpt-4" latest GPT4 model.
        /// </summary>
        Four,
        /// <summary>
        /// "gpt-4-0314", fixed GPT4 model at 03/14/2023.
        /// </summary>
        [Obsolete("Use Four0613 instead of Four0314 until 09/13/2023.")]
        Four0314,
        /// <summary>
        /// "gpt-4-0613", fixed GPT4 model at 06/13/2023.
        /// </summary>
        Four0613,
        /// <summary>
        /// "gpt-4-32k", latest GPT4 model for 32k tokens.
        /// </summary>
        Four32K,
        /// <summary>
        /// "gpt-4-32k-0314", fixed GPT4 model for 32k tokens at 03/14/2023.
        /// </summary>
        [Obsolete("Use Four32K0613 instead of Four32K0314 until 09/13/2023.")]
        Four32K0314,
        /// <summary>
        /// "gpt-4-32k-0613", fixed GPT4 model for 32k tokens at 06/13/2023.
        /// </summary>
        Four32K0613,
    }
}